src/harness/test_harness_objects.py

```
@@ -0,0 +1,324 @@
+    """
+    # Concatenate all the heartbeat request of CBSD into single request
+    heartbeat_requests = []
+    for cbsd_object_item in self.cbsd_objects.values():
+      heartbeat_requests.extend(cbsd_object_item.constructHeartbeatRequestForAllActiveGrants())
+
+    if len(heartbeat_requests):
+      heartbeat_requests_wrapped = {
+        'heartbeatRequest': heartbeat_requests
+      }
+      print heartbeat_requests_wrapped
+      # Perform heartbeat requests
+      heartbeat_responses = self.testcase._sas.Heartbeat(
+        heartbeat_requests_wrapped, self.ssl_cert, self.ssl_key)['heartbeatResponse']
+
+      for heartbeat_response in heartbeat_responses:
+        cbsd_object = self.cbsd_objects[heartbeat_response['cbsdId']]
+        grant_object = cbsd_object.grant_objects[heartbeat_response['grantId']]
+
+        # Update grant state based on the heartbeat response
+        self._mapResponseCodeToGrantState(heartbeat_response['response']['responseCode'],
+                                         grant_object)
+
+      return heartbeat_requests, heartbeat_responses
+
+  def _mapResponseCodeToGrantState(self, heartbeat_response_code, grant_object):
+    """Maps the heartbeat response and update the grant state."""
+    if heartbeat_response_code == Grant.ResponseCodes.SUCCESS.value:
+      grant_object.authorized_in_last_heartbeat = True
+    elif heartbeat_response_code == Grant.ResponseCodes.TERMINATED_GRANT.value:
+      grant_object.is_terminated = True
+    elif heartbeat_response_code == Grant.ResponseCodes.SUSPENDED_GRANT.value:
+      grant_object.authorized_in_last_heartbeat = False
+    elif heartbeat_response_code == Grant.ResponseCodes.UNSYNC_OP_PARAM.value:
+      grant_object.is_terminated = True
+      grant_object.authorized_in_last_heartbeat = False
+      logging.error('Received UNSYNC_OP_PARAM code in Heartbeat Response')
+    else:
+      logging.error('Unknown code=%s received in Heartbeat Response',heartbeat_response_code)
```

Please also raise an exception, this should never happen during testing.

---

Raised   exception incase of heartbeat response unknown error code.